### PR TITLE
Add I3-login-logout

### PR DESCRIPTION
**Pull request by vinhnd2010**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (dieunb) [inline app/controllers/sessions_controller.rb]: `"` not `'`

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -13,4 +13,29 @@ def user_params
...
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = t("user.pleaselogin")
+      redirect_to login_url
+    end
+  end
+
+  def supervisor
+    if logged_in?
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: why not `if logged_in? && !supervisor?`

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
+module SessionsHelper
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if (user_id = session[:user_id])
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: remove `()`

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: remove `()`

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      user = User.find_by(id: user_id)
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: remove `()`

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+      redirect_to root_url
+    else
+      flash[:danger] = t("invalid")
+      render "new"
```
Comment (ThuBM) [inline app/controllers/sessions_controller.rb]: :new

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
+      redirect_to root_url
+    else
+      flash[:danger] = t("invalid")
```
Comment (ThuBM) [inline app/controllers/sessions_controller.rb]: bỏ ()

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    cookies.delete :remember_token
+  end
+
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def supervisor?
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: move to model, write instance method

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    current_user.role == 1
+  end
+
+  def teacher?
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: move to model, write instance method

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (ThuBM) [inline app/controllers/sessions_controller.rb]: '1' => "1"

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -13,4 +13,29 @@ def user_params
...
 
+  def correct_user
+    @user = find_object_model User, params[:id]
+    redirect_to root_url unless current_user = @user
+  end
+
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = t("user.pleaselogin")
```
Comment (ThuBM) [inline app/controllers/users_controller.rb]: bỏ ()

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -13,4 +13,29 @@ def user_params
...
+      redirect_to login_url
+    end
+  end
+
+  def supervisor
+    if logged_in?
+      unless supervisor?
+        flash[:danger] = t("user.denied")
```
Comment (ThuBM) [inline app/controllers/users_controller.rb]: bỏ ()

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -13,4 +13,29 @@ def user_params
...
+        redirect_to root_url
+      end
+    else
+      store_location
+      flash[:danger] = t("user.pleaselogin")
```
Comment (ThuBM) [inline app/controllers/users_controller.rb]: như trên

File: app/helpers/sessions_helper.rb [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    current_user.role == 2
+  end
+
+  def redirect_back_or defaul
```
Comment (ThuBM) [inline app/helpers/sessions_helper.rb]: defaul => default

**Comment by hellovietnam93** [comment]
OK

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -13,4 +13,27 @@ def user_params
...
+    unless logged_in?
+      store_location
+      flash[:danger] = t "user.pleaselogin"
+      redirect_to login_url
+    end
+  end
+
+  def supervisor
+    if logged_in? && !supervisor?
+        flash[:danger] = t "user.denied"
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: indent

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -13,4 +13,27 @@ def user_params
...
+
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = t "user.pleaselogin"
+      redirect_to login_url
+    end
+  end
+
+  def supervisor
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: function name should `verify_supervisor`

File: app/helpers/sessions_helper.rb [hunk 16]
```diff
@@ -0,0 +1,39 @@
+module SessionsHelper
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by(id: user_id)
+    elsif user_id = cookies.signed[:user_id]
+      user = User.find_by(id: user_id)
```
Comment (dieunb) [inline app/helpers/sessions_helper.rb]: remove `()`

File: app/models/user.rb [hunk 17]
```diff
@@ -36,6 +37,50 @@ def to_param
...
+  end
+
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def supervisor?
+    current_user.role == 1
```
Comment (dieunb) [inline app/models/user.rb]: using enum rails for role

**Comment by ThuBM** [comment]
OK

File: app/controllers/sessions_controller.rb [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
+      redirect_to root_url
+    else
+      flash[:danger] = t("invalid")
```
Comment (ThuBM) [inline app/controllers/sessions_controller.rb]: bỏ ()

**Comment by ThuBM** [comment]
OK

**Pull request merged**